packages/loader/container-loader/src/deltaManager.ts

```diff
@@ -824,6 +836,22 @@ export class DeltaManager<TConnectionManager extends IConnectionManager>
             message.contents = JSON.parse(message.contents);
         }
 
+        // validate client sequence number has no gap. If there is gap, check if there were noops
+        // if there were noops, decrement the gap by number of noops and continue
+        if (this.lastClientSequenceNumber !== undefined && this.lastSubmittedClientId !== undefined && this.lastSubmittedClientId === message.clientId) {
+            const clientSeqNumGap = message.clientSequenceNumber - this.lastClientSequenceNumber - 1;
+            if (clientSeqNumGap > 0) {
+                if (this.noOpCount > 0) {
+                    this.noOpCount -= clientSeqNumGap;
+                } else {
+                    throw new Error(`gap in client sequence number :${clientSeqNumGap}`);
+                }
+            } else if (message.type === MessageType.NoOp){
```

Comment:
I do not think else-if is correct here. There could be both a gap and incoming op could be noop.
So it feels like you want to do this.noOpCount-- independently from the clientSeqNumGap > 0 check, and likely do it before clientSeqNumGap > 0 check, such that you have more precise (lower) value of this.noOpCount to validate.